### ISSUE TITLE
Memory leak when using rtl8723be on 4.18.x

### ISSUE THREAD

**Issue by coolsidd** [lwfinger/rtlwifi_new#401]
I am experiencing huge memory leak issues after switching to 4.18.x . I can confirm that these did not occur at 4.17.11. Here is the output of [kmemleak]
Also you can refer to the forum thread [here][arch forum link] for the entire history.
The bug appears to be only in the 4.18.x version of the kernel as reverting to 4.17.x and LTS both seem to fix the leak.

[kmemleak]: https://gist.github.com/coolsidd/d8a1d5addafd6a2367b68e6a6b243dc4
[arch forum link]: https://bbs.archlinux.org/viewtopic.php?pid=1808432#p1808432

**Comment by coolsidd** [comment]
Here is the output of lspci -vnn

```
00:00.0 Host bridge [0600]: Intel Corporation Xeon E3-1200 v5/E3-1500 v5/6th Gen Core Processor Host Bridge/DRAM Registers [8086:1904] (rev 08)
	Subsystem: Hewlett-Packard Company Xeon E3-1200 v5/E3-1500 v5/6th Gen Core Processor Host Bridge/DRAM Registers [103c:8136]
	Flags: bus master, fast devsel, latency 0
	Capabilities: [e0] Vendor Specific Information: Len=10 <?>
	Kernel driver in use: skl_uncore

00:02.0 VGA compatible controller [0300]: Intel Corporation Skylake GT2 [HD Graphics 520] [8086:1916] (rev 07) (prog-if 00 [VGA controller])
	Subsystem: Hewlett-Packard Company Skylake GT2 [HD Graphics 520] [103c:8136]
	Flags: bus master, fast devsel, latency 0, IRQ 125
	Memory at a0000000 (64-bit, non-prefetchable) [size=16M]
	Memory at 90000000 (64-bit, prefetchable) [size=256M]
	I/O ports at 6000 [size=64]
	[virtual] Expansion ROM at 000c0000 [disabled] [size=128K]
	Capabilities: [40] Vendor Specific Information: Len=0c <?>
	Capabilities: [70] Express Root Complex Integrated Endpoint, MSI 00
	Capabilities: [ac] MSI: Enable+ Count=1/1 Maskable- 64bit-
	Capabilities: [d0] Power Management version 2
	Capabilities: [100] Process Address Space ID (PASID)
	Capabilities: [200] Address Translation Service (ATS)
	Capabilities: [300] Page Request Interface (PRI)
	Kernel driver in use: i915
	Kernel modules: i915

00:04.0 Signal processing controller [1180]: Intel Corporation Xeon E3-1200 v5/E3-1500 v5/6th Gen Core Processor Thermal Subsystem [8086:1903] (rev 08)
	Subsystem: Hewlett-Packard Company Xeon E3-1200 v5/E3-1500 v5/6th Gen Core Processor Thermal Subsystem [103c:8136]
	Flags: bus master, fast devsel, latency 0, IRQ 16
	Memory at a1420000 (64-bit, non-prefetchable) [size=32K]
	Capabilities: [90] MSI: Enable- Count=1/1 Maskable- 64bit-
	Capabilities: [d0] Power Management version 3
	Capabilities: [e0] Vendor Specific Information: Len=0c <?>
	Kernel driver in use: proc_thermal
	Kernel modules: processor_thermal_device

00:14.0 USB controller [0c03]: Intel Corporation Sunrise Point-LP USB 3.0 xHCI Controller [8086:9d2f] (rev 21) (prog-if 30 [XHCI])
	Subsystem: Hewlett-Packard Company Sunrise Point-LP USB 3.0 xHCI Controller [103c:8136]
	Flags: bus master, medium devsel, latency 0, IRQ 123
	Memory at a1400000 (64-bit, non-prefetchable) [size=64K]
	Capabilities: [70] Power Management version 2
	Capabilities: [80] MSI: Enable+ Count=1/8 Maskable- 64bit+
	Kernel driver in use: xhci_hcd
	Kernel modules: xhci_pci

00:14.2 Signal processing controller [1180]: Intel Corporation Sunrise Point-LP Thermal subsystem [8086:9d31] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP Thermal subsystem [103c:8136]
	Flags: bus master, fast devsel, latency 0, IRQ 18
	Memory at a1432000 (64-bit, non-prefetchable) [size=4K]
	Capabilities: [50] Power Management version 3
	Capabilities: [80] MSI: Enable- Count=1/1 Maskable- 64bit-
	Kernel driver in use: intel_pch_thermal
	Kernel modules: intel_pch_thermal

00:16.0 Communication controller [0780]: Intel Corporation Sunrise Point-LP CSME HECI #1 [8086:9d3a] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP CSME HECI [103c:8136]
	Flags: bus master, fast devsel, latency 0, IRQ 124
	Memory at a1433000 (64-bit, non-prefetchable) [size=4K]
	Capabilities: [50] Power Management version 3
	Capabilities: [8c] MSI: Enable+ Count=1/1 Maskable- 64bit+
	Kernel driver in use: mei_me
	Kernel modules: mei_me

00:17.0 SATA controller [0106]: Intel Corporation Sunrise Point-LP SATA Controller [AHCI mode] [8086:9d03] (rev 21) (prog-if 01 [AHCI 1.0])
	Subsystem: Hewlett-Packard Company Sunrise Point-LP SATA Controller [AHCI mode] [103c:8136]
	Flags: bus master, 66MHz, medium devsel, latency 0, IRQ 122
	Memory at a1430000 (32-bit, non-prefetchable) [size=8K]
	Memory at a1436000 (32-bit, non-prefetchable) [size=256]
	I/O ports at 6080 [size=8]
	I/O ports at 6088 [size=4]
	I/O ports at 6060 [size=32]
	Memory at a1434000 (32-bit, non-prefetchable) [size=2K]
	Capabilities: [80] MSI: Enable+ Count=1/1 Maskable- 64bit-
	Capabilities: [70] Power Management version 3
	Capabilities: [a8] SATA HBA v1.0
	Kernel driver in use: ahci
	Kernel modules: ahci

00:1c.0 PCI bridge [0604]: Intel Corporation Sunrise Point-LP PCI Express Root Port #1 [8086:9d10] (rev f1) (prog-if 00 [Normal decode])
	Flags: bus master, fast devsel, latency 0, IRQ 16
	Bus: primary=00, secondary=01, subordinate=01, sec-latency=0
	I/O behind bridge: 00005000-00005fff [size=4K]
	Memory behind bridge: a1300000-a13fffff [size=1M]
	Prefetchable memory behind bridge: 0000000080000000-000000008fffffff [size=256M]
	Capabilities: [40] Express Root Port (Slot+), MSI 00
	Capabilities: [80] MSI: Enable- Count=1/1 Maskable- 64bit-
	Capabilities: [90] Subsystem: Hewlett-Packard Company Sunrise Point-LP PCI Express Root Port [103c:8136]
	Capabilities: [a0] Power Management version 3
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [140] Access Control Services
	Capabilities: [220] Secondary PCI Express <?>
	Kernel driver in use: pcieport

00:1c.4 PCI bridge [0604]: Intel Corporation Sunrise Point-LP PCI Express Root Port #5 [8086:9d14] (rev f1) (prog-if 00 [Normal decode])
	Flags: bus master, fast devsel, latency 0, IRQ 16
	Bus: primary=00, secondary=02, subordinate=02, sec-latency=0
	I/O behind bridge: 00004000-00004fff [size=4K]
	Memory behind bridge: a1200000-a12fffff [size=1M]
	Prefetchable memory behind bridge: 00000000a1000000-00000000a10fffff [size=1M]
	Capabilities: [40] Express Root Port (Slot+), MSI 00
	Capabilities: [80] MSI: Enable- Count=1/1 Maskable- 64bit-
	Capabilities: [90] Subsystem: Hewlett-Packard Company Sunrise Point-LP PCI Express Root Port [103c:8136]
	Capabilities: [a0] Power Management version 3
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [140] Access Control Services
	Capabilities: [220] Secondary PCI Express <?>
	Kernel driver in use: pcieport

00:1c.5 PCI bridge [0604]: Intel Corporation Sunrise Point-LP PCI Express Root Port #6 [8086:9d15] (rev f1) (prog-if 00 [Normal decode])
	Flags: bus master, fast devsel, latency 0, IRQ 17
	Bus: primary=00, secondary=03, subordinate=03, sec-latency=0
	I/O behind bridge: 00003000-00003fff [size=4K]
	Memory behind bridge: a1100000-a11fffff [size=1M]
	Prefetchable memory behind bridge: None
	Capabilities: [40] Express Root Port (Slot+), MSI 00
	Capabilities: [80] MSI: Enable- Count=1/1 Maskable- 64bit-
	Capabilities: [90] Subsystem: Hewlett-Packard Company Sunrise Point-LP PCI Express Root Port [103c:8136]
	Capabilities: [a0] Power Management version 3
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [140] Access Control Services
	Capabilities: [220] Secondary PCI Express <?>
	Kernel driver in use: pcieport

00:1f.0 ISA bridge [0601]: Intel Corporation Sunrise Point-LP LPC Controller [8086:9d48] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP LPC Controller [103c:8136]
	Flags: bus master, medium devsel, latency 0

00:1f.2 Memory controller [0580]: Intel Corporation Sunrise Point-LP PMC [8086:9d21] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP PMC [103c:8136]
	Flags: bus master, fast devsel, latency 0
	Memory at a142c000 (32-bit, non-prefetchable) [size=16K]

00:1f.3 Audio device [0403]: Intel Corporation Sunrise Point-LP HD Audio [8086:9d70] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP HD Audio [103c:8136]
	Flags: bus master, fast devsel, latency 32, IRQ 127
	Memory at a1428000 (64-bit, non-prefetchable) [size=16K]
	Memory at a1410000 (64-bit, non-prefetchable) [size=64K]
	Capabilities: [50] Power Management version 3
	Capabilities: [60] MSI: Enable+ Count=1/1 Maskable- 64bit+
	Kernel driver in use: snd_hda_intel
	Kernel modules: snd_hda_intel, snd_soc_skl

00:1f.4 SMBus [0c05]: Intel Corporation Sunrise Point-LP SMBus [8086:9d23] (rev 21)
	Subsystem: Hewlett-Packard Company Sunrise Point-LP SMBus [103c:8136]
	Flags: medium devsel, IRQ 16
	Memory at a1435000 (64-bit, non-prefetchable) [size=256]
	I/O ports at 6040 [size=32]
	Kernel driver in use: i801_smbus
	Kernel modules: i2c_i801

01:00.0 Display controller [0380]: Advanced Micro Devices, Inc. [AMD/ATI] Sun XT [Radeon HD 8670A/8670M/8690M / R5 M330 / M430 / R7 M520] [1002:6660] (rev 83)
	Subsystem: Hewlett-Packard Company Radeon R5 M330 [103c:8136]
	Flags: fast devsel, IRQ 16
	Memory at 80000000 (64-bit, prefetchable) [size=256M]
	Memory at a1300000 (64-bit, non-prefetchable) [size=256K]
	I/O ports at 5000 [size=256]
	Expansion ROM at a1340000 [disabled] [size=128K]
	Capabilities: [48] Vendor Specific Information: Len=08 <?>
	Capabilities: [50] Power Management version 3
	Capabilities: [58] Express Legacy Endpoint, MSI 00
	Capabilities: [a0] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [100] Vendor Specific Information: ID=0001 Rev=1 Len=010 <?>
	Capabilities: [150] Advanced Error Reporting
	Capabilities: [270] Secondary PCI Express <?>
	Kernel modules: amdgpu

02:00.0 Ethernet controller [0200]: Realtek Semiconductor Co., Ltd. RTL810xE PCI Express Fast Ethernet controller [10ec:8136] (rev 07)
	Subsystem: Hewlett-Packard Company RTL810xE PCI Express Fast Ethernet controller [103c:80c6]
	Flags: bus master, fast devsel, latency 0, IRQ 16
	I/O ports at 4000 [size=256]
	Memory at a1200000 (64-bit, non-prefetchable) [size=4K]
	Memory at a1000000 (64-bit, prefetchable) [size=16K]
	Capabilities: [40] Power Management version 3
	Capabilities: [50] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [70] Express Endpoint, MSI 01
	Capabilities: [b0] MSI-X: Enable+ Count=4 Masked-
	Capabilities: [d0] Vital Product Data
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [140] Virtual Channel
	Capabilities: [160] Device Serial Number 01-00-00-00-36-4c-e0-00
	Capabilities: [170] Latency Tolerance Reporting
	Kernel driver in use: r8169
	Kernel modules: r8169

03:00.0 Network controller [0280]: Realtek Semiconductor Co., Ltd. RTL8723BE PCIe Wireless Network Adapter [10ec:b723]
	Subsystem: Hewlett-Packard Company RTL8723BE PCIe Wireless Network Adapter [103c:804c]
	Flags: bus master, fast devsel, latency 0, IRQ 17
	I/O ports at 3000 [size=256]
	Memory at a1100000 (64-bit, non-prefetchable) [size=16K]
	Capabilities: [40] Power Management version 3
	Capabilities: [50] MSI: Enable- Count=1/1 Maskable- 64bit+
	Capabilities: [70] Express Endpoint, MSI 00
	Capabilities: [100] Advanced Error Reporting
	Capabilities: [140] Device Serial Number 00-23-b7-fe-ff-4c-e0-00
	Capabilities: [150] Latency Tolerance Reporting
	Capabilities: [158] L1 PM Substates
	Kernel driver in use: rtl8723be
	Kernel modules: rtl8723be
```

**Comment by coolsidd** [comment]
[Here][kmemleak] is the complete output of kmemleak without tampering / clearing anything. 
I am using the ant_sel and fwlps options since my laptop only has one antenae.
Could it be that the issue got fixed in 4.19, have you tested it on 4.18?

[kmemleak]: https://gist.github.com/d8a1d5addafd6a2367b68e6a6b243dc4

**Comment by lwfinger** [comment]
I just downloaded a 4.18.9 tarball and built that kernel. No memory leaks.

You do NOT need the fwlps option because you have only 1 antenna. In fact, I do not believe that option is needed. There was a time when the driver had a bug when letting the firmware control power save, but that was fixed a long time ago.

**Comment by lwfinger** [comment]
I have no idea what is causing the leak. That network buffer allocated in the indicated position is consumed and freed in some other location, probably by mac80211.

I used the kmemleak that is built into the kernel. It has always indicated real and false positive leaks for me. I an unaware of any false negatives.

**Comment by coolsidd** [comment]
Note that removing the module stops the leak and any subsequent scans dont show a leak. However even renabling the module causes the leak to start. I have reproduced this many times where I have removed the module, scanned added the modules and scanned. The results are alwys the same

**Comment by lwfinger** [comment]
Of course, the leak stops when you unload the module. Do you have any other wireless devices to test? A USB device would exercise the same paths through mac80211, etc.

Note that I am not running ARCH. My distro is openSUSE Tumbleweed.

**Comment by coolsidd** [comment]
Would usb tethering work? I could not find a usb wireless adapter per say.

**Comment by lwfinger** [comment]
No. You need to be running the USB adapter under Linux. E-Bay lists almost 12,000 items under the search "usb wireless adapter". Try to find a simple chip - something like RTL8187L. That will not cause any driver problems, and it works very reliably. You should pay less than $10 US.

**Comment by SmallSharky** [comment]
Ho! I have this issue in my pc too (very huge memory leak). Now I know what are doing it. Kernel 4.18.12

**Comment by lwfinger** [comment]
There has been some discussion on linux-wireless@vger.kernel.org. The problem is that mac80211 is failing to release some buffers. Although the wifi driver looks to be responsible, that is not the case.

The problem is fixed in kernel:HEAD, which will soon turn into 4.19.0.

**Comment by gsaga** [comment]
So very soon ,penguins with realtek wifi cards will be getting good signal. Thanks to people who made this possible.

**Comment by lwfinger** [comment]
https://marc.info/?l=linux-wireless&m=153829948423484&w=2

The problem was somewhere in mac80211 where it failed to deallocate an skb after processing.

**Comment by avisoftware** [comment]
I don't know if this is the right place to report,  
but i have the same issue on ubuntu 18.10 kernel 4.19. 
Thanks

**Comment by lwfinger** [comment]
This problem is not in the wireless driver, but in the wireless extensions compatibility layer. The fix is in the mainline kernel as commit 848e616e66d4592fe9afc40743d3504deb7632b4. It should have been ported to stable kernels, but is not obviously in your kernel. It should be available as 4.19.1, or so.

A workaround is to not use the compatibility routines such as iwconfig, etc. If utility iw is used, the leak does not happen.

**Comment by franmagneto** [comment]
I'm running 4.19.1 and still have the leak. I'll try to provide more information.

**Comment by lwfinger** [comment]
Until commit 848e616e66d4592fe9 propagates to stable, you will need to apply this patch to your kernel, or stop using the wext-compat network control commands.

Just as I stated almost one month ago (see above), this leak is not the driver's fault. It arises because of a bug introduced in mac80211.

**Comment by franmagneto** [comment]
Yes, I've read that already, but loading rtl8723be is the way to reproduce the issue on my system. Maybe I lack knowledge on how to do it on a more "fine tuned" approach.

I did a git clone on https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git and checkout to v4.19 here, and I can see the mentioned commit on git log. Did I miss something?

**Comment by lwfinger** [comment]
I think you would have the same leak for EVERY wireless driver that uses mac80211. Try it with a USB wifi stick.

As you have cloned the full kernel source, then you are in a perfect position to bisect between a "bad" kernel and a "good" one. I cannot do it as my system uses NetworkManager to control networks, and it only uses the nl80211 controls, not the obsolete wext variety.

**Comment by franmagneto** [comment]
I was using NetworkManager too. I've uninstalled it and now I'm using only dhcpcd with wpa_supplicant hook, and the leak still occurs.

**Comment by franmagneto** [comment]
Before this I tried to change the NetworkManager backend to iwd (I didn't know about the "new" nl80211 controls or the "old" wext, or what manager uses one or the other).

**Comment by franmagneto** [comment]
I think it have to do with this issue, because it started with 4.18.x and goes away if I unload the module that controls my wifi, but I can be wrong. If that's the case, I will be happy to help identifying what is causing it.

**Comment by lwfinger** [comment]
I cannot duplicate this issue!!!! As I have said several times, someone that does have the problem needs to bisect the mainline kernel to locate the bad commit. If you do not know how to do this, there are several tutorials available.

**Comment by franmagneto** [comment]
Ok, thanks for your response and sorry for the annoyance, I will try to locate the bad commit and I will open a new issue if I find it.

**Comment by lwfinger** [comment]
No, keep the results of the bisection in this issue. That way the most people with the problem will see the answer.

Sorry that I was rough with you, but when I have to repeat the same stuff for the 3rd or 4th time, I get angry.

**Comment by franmagneto** [comment]
No problem. I will do that.

**Comment by franmagneto** [comment]
The culprit is commit 0a9f8f0a1ba9688a9ff5f6b83c4cc1eecd2fc9f2 (rtlwifi: fix btmpinfo timeout while processing C2H_BT_INFO).

**Comment by franmagneto** [comment]
I will test master with this commit reverted now.

**Comment by franmagneto** [comment]
Reverting this commit solves the problem for me.

**Comment by lwfinger** [comment]
Thanks for the testing. After looking at that commit, the only thing I can figure is that the code is leaking an skb when rtl_c2h_fast_cmd(hw, skb) is true. If that analysis is correct, try the following (untested) patch:

diff --git a/drivers/net/wireless/realtek/rtlwifi/base.c b/drivers/net/wireless/realtek/rtlwifi/base.c
index f4122c8fdd97..3fc3b0314fce 100644
--- a/drivers/net/wireless/realtek/rtlwifi/base.c
+++ b/drivers/net/wireless/realtek/rtlwifi/base.c
@@ -2289,6 +2289,7 @@ void rtl_c2hcmd_enqueue(struct ieee80211_hw *hw, struct sk_buff *skb)
 
        if (rtl_c2h_fast_cmd(hw, skb)) {
                rtl_c2h_content_parsing(hw, skb);
+               kfree_skb(skb);
                return;
        }

Larry

**Comment by franmagneto** [comment]
It works

**Comment by lwfinger** [comment]
Thanks for testing. If you want to get credit for reporting and testing this issue, please send your full name to me at Larry.Finger@lwfinger.net. Note, putting this info in the commit message could expose your E-mail address to the Internet.

**Comment by franmagneto** [comment]
Thank you for fixing the issue!

**Comment by alambike** [comment]
Hi, any forecast when this will be available in kernel 4.19? 
Thanks a lot.

**Comment by lwfinger** [comment]
Thanks for reminding me. It appears that the ball got dropped. I just sent a message (https://marc.info/?l=linux-wireless&m=154463734306716&w=2) to the maintainer. I hope that will get some action.

**Comment by xxxtonixxx** [comment]
I think it has been accepted https://patchwork.kernel.org/patch/10687671/ :smile:

will it be in Linux 4.20?

**Comment by lwfinger** [comment]
Probably not. Although it has been accepted into wireless, I have not yet seen it pushed on to net. That needs to happen before it gets to kernel head, and 4.20 will be released just before Christmas. Note, however, that once it gets to 4.21-rcX, it will be applied to all kernels from 4.18+.

I have been just as unhappy about the delay as any of you; however, yelling at the maintainer would not have been any more productive than having one of you yell at me! :)

**Comment by wioo** [comment]
I guess it is in 4.20
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/diff/drivers/net/wireless/realtek/rtlwifi/base.c?id=v4.20&id2=v4.19

**Comment by lwfinger** [comment]
Yes, it was applied between 4.20.0-rc7 and the final 4.20.0.
